New docs (#582)
leancodepl/corelibrary: docs/cqrs/query/index.md

# Query

> Query is just an class that implements the `IQuery<TResult>` interface (there's also non-generic `IQuery` interface but it shouldn't be used directly). The only generic parameter specifies the type that the query returns when executed. It should be a DTO (because most of the time it will be serialized). Queries get the data from the system but don't modify it.

Reviewer: **a** class

## Contract

Consider the following query

```csharp
public class ProjectInfoDTO
{
    public string ProjectId { get; set; }
    public string Name { get; set; }
}

[AllowUnauthorized]

> public class FindProjectsMatchingName : IQuery<List<ProjectInfoDTO>>

Reviewer: Query's name should describe the result and means of getting it e.g. `ProjectNameById`, emphasizing the read-side of CQRS. `Find` shouldn't be used as we would like to set an example here.

It would be also worth to add a short paragraph describing naming convention.

Query handlers execute queries. They should not have any side effects but can return data back to the client. Since they can return data to the client, they don't need separate [validation] (handler can do it internally). In query handlers you don't need to operate on aggregate level (as this is read-side and is relatively DDD-free) and are allowed to perform arbitrary SQL queries.

For the above query, you can have handler like this:

```csharp
public class FindProjectsMatchingNameQH
    : IQueryHandler<FindProjectsMatchingName, List<ProjectInfoDTO>>
{
    private readonly CoreDbContext dbContext;

    public FindProjectsMatchingNameQH(CoreDbContext dbContext)
    {
        this.dbContext = dbContext;
    }

    public async Task<List<ProjectInfoDTO>> ExecuteAsync(
        HttpContext context,
        FindProjectsMatchingName query)
    {
        // Here, we use raw SQL but you are free
        // to use other mechanisms to get the data
        var filter = $"%{query.NameFilter.ToLower()}%";
        var results = await dbContext.QueryAsync<ProjectInfoDTO>(@"
            SELECT ""Id"" AS ""ProjectId"", ""Name"" FROM ""Projects""
            WHERE ""Name"" LIKE @filter",

>             new { filter });

Reviewer: This seems a bit too obscure for the base example. Maybe let's just change to getting some object by Id and simple EF+Linq?

For the above query, you can have handler like this:

```csharp
public class ProjectsQH : IQueryHandler<Projects, List<ProjectDTO>>
{
    private readonly CoreDbContext dbContext;

    public ProjectsQH(CoreDbContext dbContext)
    {
        this.dbContext = dbContext;
    }

    public async Task<List<ProjectDTO>> ExecuteAsync(
        HttpContext context,
        Projects query)
    {
        // Here, we use Entity Framework but you are free
        // to use other mechanisms to get the data
        var dbQuery = dbContext.Projects.AsQueryable();

        if (!string.IsNullOrEmpty(query.NameFilter))
        {
            dbQuery = dbQuery.Where(p => p.Name.Contains(query.NameFilter));
        }

        return await dbQuery
            .Select(p => new ProjectDTO
            {
                Id = p.Id,
                Name = p.Name
            })
            .ToListAsync(context.RequestAborted);
    }
}
```

> ## Naming conventions

Reviewer: I would opt for moving this subsection up (under Contract).
Also it would be good to make it clear that namespace is also a part of contract name (same with `Command`). Similarly to what MT does [here](https://masstransit.io/documentation/concepts/messages) (apart from that MT shouldn't be a role model for us :stuck_out_tongue: ).

# Query

> Query is just a class that implements the `IQuery<TResult>` interface (there's also non-generic `IQuery` interface but it shouldn't be used directly). The only generic parameter specifies the type that the query returns when executed. It should be a DTO (because most of the time it will be serialized). Queries get the data from the system but don't modify it.

Reviewer 2: ```suggestion
Query is a class that implements the `IQuery<TResult>` interface. The only generic parameter specifies the type that the query returns when executed. It should be a DTO (because most of the time it will be serialized). Queries get the data from the system but don't modify it.
```

> (there's also non-generic `IQuery` interface but it shouldn't be used directly)
This should be a remark somewhere down below. This is too detailed for intro paragraph.

## Packages

| Package | Link | Application in section |
| --- | ----------- | ----------- |
| LeanCode.Contracts | [![NuGet version (LeanCode.Contracts)](https://img.shields.io/nuget/vpre/LeanCode.Contracts.svg?style=flat-square&logo=nuget)](https://www.nuget.org/packages/LeanCode.Contracts/2.0.0-preview.3/) | `IQuery` |
| LeanCode.CQRS.Execution | [![NuGet version (LeanCode.CQRS.Execution)](https://img.shields.io/nuget/vpre/LeanCode.CQRS.Execution.svg?style=flat-square&logo=nuget)](https://www.nuget.org/packages/LeanCode.CQRS.Execution/8.0.2260-preview/) | `IQueryHandler` |

## Contract

> Consider the query that finds all projects that match the name filter. It may be called anonymously and returns a list of `ProjectDTO`s (we use a `List` instead of a `IList` or `IReadOnlyList` because of the DTO constraint; `List` is more DTO-ish than any interface):

Reviewer 2: ```suggestion
Consider the query that finds all projects that match the name filter. It may be called anonymously and returns a list of `ProjectDTO`s:
```

> (we use a `List` instead of a `IList` or `IReadOnlyList` because of the DTO constraint; `List` is more DTO-ish than any interface):

Same as above - this probably should be a note somewhere below, not here directly.

```csharp
namespace ExampleApp.Contracts.Projects;

[AllowUnauthorized]
public class AllProjects : IQuery<List<ProjectDTO>>
{
    public string? NameFilter { get; set; }
}

public class ProjectDTO
{
    public string Id { get; set; }
    public string Name { get; set; }
}
```

## Naming conventions

Queries are designed to retrieve information without altering the system's state. To maintain a clear and consistent naming convention, queries should possess names that directly indicate the type of information being requested, including the namespace as part of the contract. An effective approach is to use descriptive nouns or noun phrases within the designated namespace, exemplified by names like:

* `ExampleApp.Contracts.Projects.AllProjects`
* `ExampleApp.Contracts.Projects.ProjectById`
* `ExampleApp.Contracts.Employees.EmployeesInAssignment`

Query handlers should similarly be named in alignment with the associated query, appending the QH suffix within the respective namespace structure.

## Handler

> Query handlers execute queries. They should not have any side effects but can return data back to the client. Since they can return data to the client, they don't need separate [validation] (handler can do it internally). In query handlers you don't need to operate on aggregate level (as this is read-side and is relatively DDD-free) and are allowed to perform arbitrary SQL queries.

Reviewer 2: ```suggestion
Query handlers execute queries. They should not have any side effects but can return data back to the client. Since they can return data to the client, they don't need separate [validation] (it is rare to need validation; if it is needed, handlers can do it internally). In query handlers you don't need to operate on aggregate level (as this is read-side and is relatively DDD-free) and are allowed to perform arbitrary SQL queries.
```
